### DISCUSSION TITLE
Automatizar a criação de iconfont

### DISCUSSION THREAD

**Discussion by felipefialho** [frontendbr/forum#10] [Perguntas e Respostas]
Pessoal, estão usando algum plugin de Grunt, Gulp ou até PostCSS para automatizar essa tarefa?

Ainda faço tudo manualmente:
- Vou no site do Icomoon
- Gero o pacote
- Salvo o json

E em toda modificação preciso substituir o CSS, as fontes e o json novamente. Meio chato.

Que solução estão usando?

**Comment by iamgbayer** [comment]
Passo por esse processo também, estou acompanhando a issue.

**Comment by napalmdev** [comment]
Aqui o mesmo processo chato hahahah =/

**Comment by diogomoretti** [comment]
Esse processo chato também, porém, não vejo uma automação para isso!

**Comment by felipefialho** [comment]
Já vi algumas pessoas falando que usam plugin para automatizar esse processo, inclusive com libs do icomoon.

Queria muito uma solução para isso.

**Comment by matheuslc** [comment]
Não daria para fazer um plugin do gulp para automatizar isso? Tipo passa um array com os icones e ele faz o resto.

**Comment by abnersajr** [comment]
Uso o Iconmoon app também. Tenho uma conta cadastrada, crio um projeto cada vez. 
Sempre que preciso mecher em algo está salvo no meu login.
Sempre que adiciono um icone novo, movo ele para o fim da lista, para que os unicodes não baguncem, caso baguncem mando resetar a partir do primeiro que eu já tinha e então só adiciono as classes novas. Logo só troco os font files.

**Comment by abnersajr** [comment]
Dando uma pesquisada encontrei esse pacote para quem usa gulp.
https://github.com/nfroidure/gulp-iconfont

**Comment by vitor-mendrone-deel** [comment]
Esse problema surgiu ontem pra mim. Normalmente usava soluções já prontas tipo o Font Awesome ou ícones em SVG. Vou acompanhar o tópico pra ver o que a galera ta usando de solução :smiley:

**Comment by idmarjr** [comment]
Eu só uso o Font Awesome como iconfont pela facilidade.

Mas sempre que preciso de ícones personalizados faço isso usando SVG Sprite. É mais leve, mais confiável e menos gambiarra (Porque honestamente, iconfonts é uma baita gambiarra que virou padrão.)
- https://css-tricks.com/svg-sprites-use-better-icon-fonts/
- https://24ways.org/2014/an-overview-of-svg-sprite-creation-techniques/
- http://webdesign.tutsplus.com/tutorials/how-to-implement-cross-browser-svg-sprites--cms-22427

Mas infelizmente crio isso manualmente também. :disappointed:

**Comment by willycamargo** [comment]
Aqui o processo é paracido @LFeh , o rapaz que trabalhava comigo criou uma task bacana no gulp pra resolver isso ai:

**gulp/fonticon.js**

```
/**
 * dependencies
 */
import del from 'del'
import flatten from 'gulp-flatten'
import gulp from 'gulp'
import runSequence from 'run-sequence'
import unzip from 'gulp-unzip'

/**
 * tasks
 */
gulp.task('unzip:fonticon', () =>
  gulp.src('./src/font/fonticon.zip')
    .pipe(unzip())
    .pipe(gulp.dest('./src/font/fonticon'))
)

gulp.task('copy:fonticon', () =>
  gulp.src('./src/font/fonticon/fonts/**')
    .pipe(flatten())
    .pipe(gulp.dest('./src/font/'))
)

gulp.task('clean:fonticon-fonts', () =>
  del([ './src/font/fonticon-*' ])
)

gulp.task('clean:fonticon-temp', () =>
  del([ './src/font/fonticon', './src/font/fonticon.zip' ])
)

gulp.task('fonticon', () =>
  runSequence('unzip:fonticon', 'clean:fonticon-fonts', 'copy:fonticon', 'clean:fonticon-temp', 'font')
)
```

O processo de gerar o zip e o json no site do icomoon ainda é igual, mas depois de salvo, apenas rodo `gulp fonticon` que o trabalho monótono é feito :smile:

**Comment by rickbenetti** [comment]
Concordo com o @idmarjr usar o ícone como fonte é muita gambiarra, e estou começando a estudar as dicas do @willianjusten  que inclusive são muito boas no blog.

Segue o link sobre ícones:
- http://willianjusten.com.br/sistemas-de-icones-em-svg/

Agora automatizar isso, com SVG ao menos seria genial, mas precisaria manjar de JS ao menos eu imagino pra gerar um plugin que abra a lista de SVG, pegue o código necessário e junte eles em um só criando um sprite de SVG, não sei se isso seria possível.

**Comment by rickbenetti** [comment]
Ah sim lembrei de um repositório que sigo que pra ícones tem em png e svg dependendo do tipo de navegador ele irá carregar e com o **SASS**, **LESS** e **Stylus** é possível criar functions pra gerar os que são usados senão em engano.
- https://github.com/adamfairhead/webicons

**Comment by victorserpac** [comment]
Uso SVG sprites há mais de um ano e nunca mais me preocupei com font-icons. Uso o [gulp-svgstore](https://github.com/w0rm/gulp-svgstore) junto com o [gulp-svgmin](https://github.com/ben-eb/gulp-svgmin) para gerar o sprite e consigo otimizar esse processo, só largo os SVG dentro da pasta de icons e o gulp faz todo o trabalho pra mim.

**Comment by felipefialho** [comment]
Já pensei nessa solução @victorserpac. Consegue enumerar seu passo a passo para a gente ter uma ideia?

**Comment by vitorbritto** [comment]
A depender do projeto, utilizo o Icomoon e seu processo braçal.

Senão, utilizo o mesmo processo do @victorserpac (via Grunt).
- https://github.com/FWeinb/grunt-svgstore
- https://github.com/sindresorhus/grunt-svgmin
- https://github.com/dbushell/grunt-svg2png

**ou**

o [Grunticon](http://www.grunticon.com/).

**Comment by victorserpac** [comment]
Como falei antes eu uso o [gulp-svgstore](https://www.npmjs.com/package/gulp-svgstore) e o [gulp-svgmin](https://www.npmjs.com/package/gulp-svgmin). No meu `gulpfile.js` minha task fica dessa maneira:

```
var gulp = require('gulp');
var rename = require('gulp-rename');
var path = require('path');
var svgstore = require('gulp-svgstore');
var svgmin = require('gulp-svgmin');

gulp.task('svgstore', function () {
    return gulp
        .src('assets/images/icons/*.svg')
        .pipe(svgmin())
        .pipe(svgstore({ inlineSvg: true }))
        .pipe(rename(function(filepath) {
            filepath.basename = 'icons'
        }))
        .pipe(gulp.dest('dist/images/'));
});
```

Depois de executar a task (`gulp svgstore`) o arquivo `icons.svg` é criado. No meu caso eu importo ele com o php mesmo, mas é possível injetar o svg com o [gulp-inject](https://github.com/klei/gulp-inject).

Meu HTML fica mais ou menos assim:

```
<!DOCTYPE html>
<html lang="pt-BR">
<head></head>
<body>
    <?php
    $svg_sprites = 'dist/images/icons.svg';

    if( file_exists($svg_sprites) ):
    ?>
        <div style="height: 0; width: 0; position: absolute; left:-9999px; visibility: hidden">
            <?php include($svg_sprites); ?>
        </div>
    <?php endif; ?>

    ...

</body>
</html>
```

Agora é só usar `<svg><use xlink:href="#svgfilename"></use></svg>` mudando _svgfilename_ pelo nome do arquivo sem a extensão .svg. É possível aplicar estilo na tag svg normalmente como feito em svg normal, ficando assim muito simples para mudar tamanho, cor e tudo mais.

Basicamente esse é o meu workflow com SVG Sprites, gosto muito do [Noun Project](https://thenounproject.com) para pegar os SVGs, quase todos os icons que preciso encontro lá.

É isso, qualquer dúvida estou à disposição e aceito sugestões também para alguma melhoria :D

**Comment by felipefialho** [comment]
Sensacional @victorserpac

**Comment by rickbenetti** [comment]
Animal isso, sensacional+1 @victorserpac

**Comment by vitor-mendrone-deel** [comment]
@victorserpac cara, que foda!

**Comment by ghost** [comment]
@victorserpac, nunca tinha usado SVG sprites, muito obrigado por compartilhar cara!

**Comment by willianjusten** [comment]
Similar ao @victorserpac, eu também trabalho com Sprites SVG. No Globoesporte, eu precisei criar um conjunto de sprites dos escudos de vários times e para isso eu criei dois projetinhos bem básicos, mas que fazem compressão/otimização nas melhores práticas, além de criar os sprites e também fallbacks.

Se alguém se interessar, eu deixei os projetos separados:
- [SVG Gulp Optimizer](https://github.com/willianjusten/svg-gulp-optimizer)
- [SVG Grunt Optimizer](https://github.com/willianjusten/svg-grunt-optimizer)

**Ps.:** esses projetinhos foram bem básicos para solucionar basicamente minhas necessidades no Globoesporte, mas podem ser estendidos e aceito N contribuições =)

Eu também fiz dois posts explicando processos de criação de ícones/sprites e diferentes utilizações no SVG (com o `<use>` e com o Fragment Identifiers), seguem os links também:
- [Sistemas de Ícones em SVG](http://willianjusten.com.br/sistemas-de-icones-em-svg/)
- [Usando SVG Sprites](http://willianjusten.com.br/usando-svg-sprites/)

**Comment by felipefialho** [comment]
Valeu pela contribuição @willianjusten. Vou testar SVG ao invés de iconfont.

**Comment by vitor-mendrone-deel** [comment]
Vocês viram [isso](http://dev.edenspiekermann.com/2016/02/10/introducing-spritesh/) @LFeh @willianjusten ?

É basicamente um bash script que cria um spritesheet a partir de SVGs.

**Comment by vitorbritto** [comment]
Shell Script :heart:

**Comment by felipefialho** [comment]
Não tinha visto @Mendrone. Cara, preciso migrar pra Gulp o quanto antes, parece ser mais tranquila a utilização desses plugins de automação de gráficos.

**Comment by vitorbritto** [comment]
@LFeh comecei hoje a utilizar p gulp em todos os meus projetos correntes. Migrei tudo no final de semana. Bower também foi embora. Agora uso JSPM. 

O Grunt foi um ótimo companheiro de trabalho, é uma ferramenta excepcional! Quem me conhece, sabe o quanto aprecio a ferramenta. Até montei uma configuração performática, mas não foi o suficiente para execução.

**Comment by willianjusten** [comment]
Gulp é muito melhor para tudo <3
Já usei de tudo, mas Gulp tá no coração =)

**Comment by idmarjr** [comment]
Github trocou a iconfont por SVG:
https://github.com/blog/2112-delivering-octicons-with-svg

**Comment by felipefialho** [comment]
Só completando. Passei a usar SVG para os icons usando o Gulp, após esse [commit do @zehfernandes](https://github.com/frontendbr/eventos/commit/76c4c2d24ffbf6e0c6bf91bedb7dbe34263e4c12).

Já implementei no [Kratos Boilerplate](https://github.com/LFeh/kratos-boilerplate).

**Comment by diegodap** [comment]
Opa,
Aproveitando o tópico, vocês recebem layout no Sketch? 
Tem problemas as vezes na hora de exportar o svg ?

Digo na tela ele estar de uma forma, e quando exporta aparece desconfigurado no navegador.

**Comment by GIlbertoDomingues** [comment]
Pessoal, 
estou implementando a solução do @victorserpac, aqui deu quase certo, só na merda do IE11 que não esta rolando, queria saber se estão usando algum fallback ou hack para funcionar no ie.

Se não, conhecem alguma outra forma de automatizar essa tarefa, que funcione no >IE10 ?

**Comment by LuanDc** [comment]
Aproveitando o espaço, sei que a Issue é antiga, mas prefiro não criar uma nova e ajudar a manter a organização do fórum. 

Preciso converter ícones SVG em font. Existe uma maneira de automatizar esse processo pelo Webpack? Ou apenas pelo IconMoon?

**Comment by simaodeveloper** [comment]
@LuanDc ainda não precisei fazer nada parecido, mas com uma pesquisa rápida no google, achei esse loader de webpack, https://github.com/jeerbl/webfonts-loader, e esse plugin, https://github.com/jantimon/iconfont-webpack-plugin, da uma olhada e veja se te serve bem e nos conte sua experiência.

**Comment by portothree** [comment]
[off-topic] que saudade do gulp hahahaha

**Comment by eliseumds** [comment]
Puts, é coisa chata mesmo. O nosso designer salva os SVGs numa pasta específica, e depois executa um NPM script que gera o arquivo `.woff` (suportado por todos os browsers) e um componente React+TypeScript com os nomes dos ícones baseados nos nomes dos arquivos SVG, por exemplo:

```
src/
  fonts/
    icons/
      star.svg
      trash.svg
    minha-outra-fonte/
      blabla.svg
```

Isso gera os arquivos `icons.woff` e `minha-outra-fonte.woff`. Há um outro script que gera um componente *Icon.tsx*:

```tsx
import Text, { TextProps } from './Text';

const ICON_NAME_TO_CHAR_CODE = {
  'star': 442,
  'trash': 534,
};

type Props = TextProps & {
  name: keyof typeof ICON_NAME_TO_CHAR_CODE;
};

export default function Icon(props: Props) {
  const { name, ...otherProps } = props;
  const charCode = ICON_NAME_TO_CHAR_CODE[name];

  // The className "font-icon" is a global one created by <GlobalStylesWrapper>
  return (
    <Text className="font-icon" {...otherProps}>
      {String.fromCharCode(charCode)}
    </Text>
  );
}
```

Gerar o componente é fácil (basicamente um string template). O script que eu criei pra gerar a fonte com unicode characters determinísticos (você pode alterar o SVG, mas se o nome do arquivo continuar o mesmo, não há problema): https://gist.github.com/eliseumds/b17bbb562795489933b4a5888369227d

**Comment by LuanDc** [comment]
Devido ao prazo eu fui pelo caminho mais simples (iconmoon). Mas quando me derem um tempo irei refatorar e automatizar esse processo. vllw galera :D
